domain/status/service/storage_test.go

```diff
@@ -0,0 +1,169 @@
+		Status: corestatus.Attached,
+	})
+	c.Assert(err, tc.ErrorIs, storageerrors.FilesystemNotFound)
+}
+
+func (s *storageServiceSuite) TestSetFilesystemStatusInvalidStatus(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	err := s.service.SetFilesystemStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status: "invalid",
+	})
+	c.Assert(err, tc.ErrorMatches, `.*unknown filesystem status "invalid"`)
+
+	err = s.service.SetFilesystemStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status: corestatus.Allocating,
+	})
+	c.Assert(err, tc.ErrorMatches, `.*unknown filesystem status "allocating"`)
+}
+
+func (s *storageServiceSuite) TestSetVolumeStatus(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	now := time.Now()
+
+	volumeUUID := storagetesting.GenVolumeUUID(c)
+	s.state.EXPECT().GetVolumeUUIDByID(gomock.Any(), "666").Return(volumeUUID, nil)
+	s.state.EXPECT().SetVolumeStatus(gomock.Any(), volumeUUID, status.StatusInfo[status.StorageVolumeStatusType]{
+		Status:  status.StorageVolumeStatusTypeAttached,
+		Message: "doink",
+		Data:    []byte(`{"foo":"bar"}`),
+		Since:   &now,
+	})
+
+	err := s.service.SetVolumeStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status:  corestatus.Attached,
+		Message: "doink",
+		Data:    map[string]any{"foo": "bar"},
+		Since:   &now,
+	})
+	c.Assert(err, tc.ErrorIsNil)
+
+	c.Check(s.statusHistory.records, tc.DeepEquals, []statusHistoryRecord{{
+		ns: statushistory.Namespace{Kind: corestatus.KindVolume, ID: volumeUUID.String()},
+		s: corestatus.StatusInfo{
+			Status:  corestatus.Attached,
+			Message: "doink",
+			Data:    map[string]any{"foo": "bar"},
+			Since:   &now,
+		},
+	}})
+}
+
+func (s *storageServiceSuite) TestSetVolumeStatusNotFound(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().GetVolumeUUIDByID(gomock.Any(), "666").Return("", storageerrors.VolumeNotFound)
+
+	err := s.service.SetVolumeStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status: corestatus.Attached,
+	})
+	c.Assert(err, tc.ErrorIs, storageerrors.VolumeNotFound)
+}
+
+func (s *storageServiceSuite) TestSetVolumeStatusInvalidStatus(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	err := s.service.SetVolumeStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status: "invalid",
+	})
+	c.Assert(err, tc.ErrorMatches, `.*unknown volume status "invalid"`)
+
+	err = s.service.SetVolumeStatus(c.Context(), "666", corestatus.StatusInfo{
+		Status: corestatus.Allocating,
+	})
+	c.Assert(err, tc.ErrorMatches, `.*unknown volume status "allocating"`)
+}
+
+func (s *storageServiceSuite) setupMocks(c *tc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.state = NewMockState(ctrl)
+	s.statusHistory = &statusHistoryRecorder{}
+
+	s.service = NewService(
+		s.state,
+		modeltesting.GenModelUUID(c),
+		s.statusHistory,
+		func() (StatusHistoryReader, error) {
+			return nil, errors.Errorf("status history reader not available")
+		},
+		clock.WallClock,
+		loggertesting.WrapCheckLog(c),
+	)
+
```

Comment:
In future (don't need to do it for this PR), but can we please add clean up functions that nil out shared variables.
```go
c.Cleanup(func() {
    s.state = nil
    s.statusHistory = nil
    s.service = nil
})
```

Reply:
Fixed here and updated other existing test in this package.